modules/audio/src/main/native/android/dalvik/DalvikAudioService.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2020, Gluon
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL GLUON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.gluonhq.helloandroid;
+
+import android.media.AudioAttributes;
+import android.media.AudioManager;
+import android.media.MediaPlayer;
+import android.media.SoundPool;
+
+import com.gluonhq.attach.audio.Audio;
+import com.gluonhq.attach.storage.StorageService;
```

Comment:
At the Dalvik side, these two classes won't be found.
For Storage, you could use directly the Android implementation, either from here: https://github.com/gluonhq/attach/blob/master/modules/storage/src/main/java/com/gluonhq/attach/storage/impl/AndroidStorageService.java#L55 or https://github.com/gluonhq/substrate/blob/master/src/main/resources/native/android/android_project/app/src/main/java/com/gluonhq/helloandroid/MainActivity.java#L106 that is used to define `user.home` in the first place, but then you will need to pass `Activity` to the constructor.

For `Audio` is more complex. You are creating an object (`DalvikMusic` or `DalvikSound`) in one VM and sending it to another. That won't work. You could define `Audio` again in the Dalvik side, so you can keep those two implementations. Then you will have to wire up the interface methods, so when you call `play` from the Java/GraalVM side, it ultimately calls `Audio::play` in the Dalvik side, where the audio object exists. Best to check the Video service for reference (i.e. `play`: https://github.com/gluonhq/attach/blob/master/modules/video/src/main/java/com/gluonhq/attach/video/impl/AndroidVideoService.java#L81)

Reply:
I see, thanks, I have a clearer idea now how to wire these. Will update

Reply:
Please see updated version, which passes around `int audioId` instead of objects.